File: Scheduling/roundrobin.py
import pandas as pd
import operator

class Process:
    def __init__(self):
        self.pid=input("Enter Process ID : ")
        self.bt=int(input("Enter Burst time : "))
    def get_wt(self,x):
        self.wt=x
    def get_tat(self,x):
        self.tat=x
        

def wait_time(p,qtm,n):
    rbt=[]
    t=0
    for item in p:
        rbt.append(item.bt)
    
    while(1):
        finished=True
        for i in range(0,n):
            if rbt[i]>0:
                finished=False
            
                if rbt[i]>qtm:
                    t+=qtm
                    rbt[i]-=qtm
                    
                else:
                    t=t+rbt[i]
                    wait=t-p[i].bt
                    p[i].get_wt(wait)
                    rbt[i]=0
        
        
        if finished==True:
            break
            

def turnaround_time(p,n):
    for i in range(0,n):
        turn=p[i].wt+p[i].bt
        p[i].get_tat(turn)
                
                
PID=[]
BT=[]
WT=[]
TAT=[]
n = int(input("No:of processes ? "))
p = []
for i in range(0, n):
    p.append(Process())
    PID.append(p[i].pid)
    BT.append(p[i].bt)
qtm=int(input("Time Quantum ? "))
wait_time(p,qtm,n)
turnaround_time(p,n)

for item in p:
    WT.append(item.wt)
    TAT.append(item.tat)

d={'Process #':PID,'Burst Time':BT,'Turn-around Time':TAT,'Waiting Time':WT}
ptable=pd.DataFrame(d)
s=ptable.sort_values(by=['Waiting Time','Turn-around Time'])

    
print("\n\nRound Robin Scheduling")
l=s['Process #'].tolist()
for item in l:
    print("< {} >".format(item),end=" ")
    
display(ptable)
avw=float(sum(WT)/n)
avt=float(sum(TAT)/n)
print("Average Turnaround Time : ",avt)
print("Average Waiting Time : ",avw)

File: Scheduling/C++ Implementation/roundrobin.cpp
#include<iostream>
#include<numeric>

using namespace std;
struct Process{
    string pid;
    int bt;
    int wt;
    int tat;
};



void time_calc(int qtm,Process P[],int n)
{
    int rbt[10];
    int t = 0;
    int i;

    for(i =0;i<n;i++){

        rbt[i] = P[i].bt;

    }

    while(true){
        bool finished = true;

        for(i=0;i<n;i++){
            
            if(rbt[i] > 0){
                finished = false;

                if(rbt[i] > qtm){
                    rbt[i] -= qtm;
                    t += qtm;
                }

                else{
                    t = t + rbt[i];
                    P[i].wt = t - P[i].bt;
                    
                    P[i].tat = P[i].wt + P[i].bt;
                    
                    rbt[i] = 0;
                }
            }
        }

        if (finished==true){
        break;
        }

    }


}



int main()
{
    Process P[10];

    int i,n;
    

    std::cout << "\nEnter the number of Processes";
    std::cin >> n;

    for(i=0;i<n;i++)
    {
        std::cout << "\nEnter Process ID ";
        cin >> P[i].pid;
        std::cout << "\nEnter Burst Time";
        cin >> P[i].bt;
    }

    int qtm;
    cout << "\nEnter Time QTM";
    cin >> qtm;

    time_calc(qtm,P,n);


    

    


    std::cout <<"\nPID\t" << "BT\t" << "WT\t" << "TAT";
    float avg_wt =0 , avg_tat = 0;

    for(i=0;i<n;i++){
        std::cout <<'\n'<<P[i].pid<<'\t'<<P[i].bt<<'\t'<<P[i].wt<<'\t'<<P[i].tat<<'\n';
        avg_wt += P[i].wt;
        avg_tat += P[i].tat;
    }

    std::cout<<"\nAverage Waiting Time : "<<avg_wt/n;
    std::cout<<"\nAverage Turnaround Time : "<<avg_tat/n<<'\n';
    
    return 0;
}